### PR TITLE
tilt: clean up synclet clients and tunnels

### PR DESCRIPTION
**Pull request by landism**

### CODE REVIEW

File: internal/engine/build_and_deployer_test.go [hunk 1]
```diff
@@ -353,6 +353,23 @@ RUN ["go", "install", "github.com/windmilleng/sancho"]`,
...
+	_, err := f.bd.BuildAndDeploy(f.ctx, SanchoManifest, NewBuildState(alreadyBuilt))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if f.sCli.ClosedCount != 1 {
+		t.Errorf("Expected 1 synclet client close, actual: %d", f.sCli.ClosedCount)
+	}
+
+=======
```
Comment (nicks) [inline internal/engine/build_and_deployer_test.go]: ?

File: internal/engine/build_and_deployer.go [hunk 2]
```diff
@@ -50,13 +54,27 @@ func NewCompositeBuildAndDeployer(builders BuildOrder, shouldFallBack FallbackTe
...
+
 func (composite *CompositeBuildAndDeployer) BuildAndDeploy(ctx context.Context, manifest model.Manifest, currentState BuildState) (BuildResult, error) {
 	var lastErr error
 	for _, builder := range composite.builders {
 		br, err := builder.BuildAndDeploy(ctx, manifest, currentState)
 		if err == nil {
 			// TODO(maia): maybe this only needs to be called after certain builds?
 			// I.e. should be called after image build but not after a successful container build?
+			if currentState.LastImage() != nil && br.Image != currentState.LastImage() {
+				composite.forgetImage(ctx, currentState.LastImage())
```
Comment (nicks) [inline internal/engine/build_and_deployer.go]: it seems weird that this is a separate api...? could we make this part of PostProcessBuild?

File: internal/engine/synclet_build_and_deployer.go [hunk 3]
```diff
@@ -87,6 +87,22 @@ func (sbd *SyncletBuildAndDeployer) deployInfoForImageBlocking(ctx context.Conte
 	return deployInfo, ok
 }
 
+func (sbd *SyncletBuildAndDeployer) forgetImage(ctx context.Context, img reference.NamedTagged) error {
+	sbd.deployInfoMu.Lock()
+	defer sbd.deployInfoMu.Unlock()
+
+	imgNameAndTag := docker.ToImgNameAndTag(img)
+
+	deployInfo, ok := sbd.deployInfo[imgNameAndTag]
```
Comment (nicks) [inline internal/engine/synclet_build_and_deployer.go]: are we concerned about the cases where an image is deployed more than once, to different pods?

File: internal/engine/build_and_deployer_test.go (hunk 1 above)
Comment (landism) [reply internal/engine/build_and_deployer_test.go]: sorry, sloppy merge

File: internal/engine/build_and_deployer.go (hunk 2 above)
Comment (landism) [reply internal/engine/build_and_deployer.go]: good call

File: internal/engine/synclet_build_and_deployer.go (hunk 3 above)
Comment (landism) [reply internal/engine/synclet_build_and_deployer.go]: I think we should be, but that's also broken independently of this change (i.e., we're already keying podid + synclet client by image)

File: internal/engine/build_and_deployer.go [hunk 4]
```diff
@@ -23,6 +26,8 @@ type BuildAndDeployer interface {
 	// In general, we'll store this info ON the BuildAndDeployer that needs it.
 	// Each implementation of PostProcessBuild is responsible for executing long-running steps async.
 	PostProcessBuild(ctx context.Context, result BuildResult)
+
+	forgetImage(ctx context.Context, image reference.NamedTagged) error
```
Comment (maiamcc) [inline internal/engine/build_and_deployer.go]: any reason this one is private and the rest are public?

File: internal/engine/build_and_deployer.go [hunk 5]
```diff
@@ -50,13 +55,29 @@ func NewCompositeBuildAndDeployer(builders BuildOrder, shouldFallBack FallbackTe
 	}
 }
 
+func (composite *CompositeBuildAndDeployer) forgetImage(ctx context.Context, image reference.NamedTagged) error {
+	for _, builder := range composite.builders {
+		fmt.Printf("forgetting on %T\n", builder)
```
Comment (maiamcc) [inline internal/engine/build_and_deployer.go]: stray print? (/should use logger?)

File: internal/engine/build_and_deployer.go (hunk 2 above)
Comment (maiamcc) [reply internal/engine/build_and_deployer.go]: ++

File: internal/engine/sidecar_synclet_manager.go [hunk 6]
```diff
@@ -66,6 +84,22 @@ func (ssm SidecarSyncletManager) ClientForPod(ctx context.Context, podID k8s.Pod
 	return client, nil
 }
 
+func (ssm SidecarSyncletManager) ForgetPod(ctx context.Context, podID k8s.PodID) error {
+	ssm.mutex.Lock()
+	defer ssm.mutex.Unlock()
+
+	client, ok := ssm.clients[podID]
+	if !ok {
+		fmt.Printf("no client for pod '%s', so nothing to close!\n", podID)
```
Comment (maiamcc) [inline internal/engine/sidecar_synclet_manager.go]: use logger

File: internal/engine/synclet_build_and_deployer.go (hunk 3 above)
Comment (maiamcc) [reply internal/engine/synclet_build_and_deployer.go]: 👀 
uhhh.
well the answer to your question is "no". but maybe that's not the RIGHT answer...

**Comment by maiamcc** [comment]
a lot of `fmt.Printf`'s around that should probably either be deleted or be `logger` calls...

**Comment by landism** [comment]
> a lot of `fmt.Printf`'s around that should probably either be deleted or be `logger` calls...

I think I got all of them in the commit last night. Sorry the sloppy PR.

**Pull request merged**